<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ TxRegInfra2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TxRegInfra2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/TxRegInfra2.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/shims.html">shims in TxRegInfra2: dealing with heterogeneous annotation native to different genomic assay archives</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vjcitn/TxRegInfra2">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="RaggedMongoExpt.html">RaggedMongoExpt()</a></code> </p>
        </td>
        <td><p>bind colData to a mongo-based ragged-experiment incubator</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="URL_txregInAWS.html">URL_txregInAWS()</a></code> </p>
        </td>
        <td><p>return mongodb URL for working mongo server</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="URL_txregLocal.html">URL_txregLocal()</a></code> </p>
        </td>
        <td><p>local mongodb txregnet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addsyms.html">addsyms()</a></code> </p>
        </td>
        <td><p>add gene symbols to a GRanges that uses ensembl identifiers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="basicColData.html">basicColData</a></code> </p>
        </td>
        <td><p>basicColData: metadata about a small collection of files for demonstrating TxRegInfra2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="basicColData.tiny.html">basicColData.tiny</a></code> </p>
        </td>
        <td><p>basicColData.tiny: metadata about a tiny collection of files for demonstrating TxRegInfra2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="basicFormatter.html">basicFormatter()</a></code> </p>
        </td>
        <td><p>operate on a character vector to derive a DataFrame, splitting on a tokenand retrieving first and last split fragments as 'base' and 'type' fields</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="demo_eQTL_granges.html">demo_eQTL_granges</a></code> </p>
        </td>
        <td><p>a GRanges instance with eQTL metadata returned by `sbov`</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="demo_fimo_granges.html">demo_fimo_granges</a></code> </p>
        </td>
        <td><p>a list of GRanges instances with TF FIMO scores returned by `fimo_granges`</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dgf_meta.html">dgf_meta</a></code> </p>
        </td>
        <td><p>dgf_meta: metadata about a small collection of bed files for demonstrating TxRegInfra2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dnmeta.html">dnmeta</a></code> </p>
        </td>
        <td><p>metadata about DNaseI hotspots from ENCODE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dnmetaApp.html">dnmetaApp()</a></code> </p>
        </td>
        <td><p>support search of hotspot/peak data from ENCODE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDocumentFields.html">getDocumentFields()</a></code> </p>
        </td>
        <td><p>determine the fields present in a txregnet document</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFieldNames.html">getFieldNames()</a></code> </p>
        </td>
        <td><p>get names of fields in a collection in remote txregnet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="grConverter.html">grConverter()</a></code> </p>
        </td>
        <td><p>convert a GRanges to a JSON query for mongodb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importBedToMongo.html">importBedToMongo()</a></code> </p>
        </td>
        <td><p>arrange import to mongo using mongoimport, setting up type and fields appropriately</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listAllCollections.html">listAllCollections()</a></code> </p>
        </td>
        <td><p>list all collections in a database, using command-line interface</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeAggregator.html">makeAggregator()</a></code> </p>
        </td>
        <td><p>generate JSON to aggregate (counting records, and, by default, averaging
a given variable) within a collection</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeColData.html">makeColData()</a></code> </p>
        </td>
        <td><p>generate a colData component corresponding to a mongodb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGRConverterList.html">makeGRConverterList()</a></code> </p>
        </td>
        <td><p>generate a list of GRanges to JSON for queries to mongo</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oldbasicColData.html">oldbasicColData</a></code> </p>
        </td>
        <td><p>oldbasicColData: metadata about a small collection of files for demonstrating TxRegInfra2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prune_to_server_line.html">prune_to_server_line()</a></code> </p>
        </td>
        <td><p>utility to prune away header lines on mongodb command line query</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ragged41FP.html">ragged41FP</a></code> </p>
        </td>
        <td><p>ragged41FP: A RaggedExperiment instance with digital genomic footprints over the coding region of ORMDL3</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sbov.html">sbov()</a></code> </p>
        </td>
        <td><p>prototype of subsetter for mongo resource</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sbov_output_FP.html">sbov_output_FP</a></code> </p>
        </td>
        <td><p>a GRanges instance with digital genomic footprint metadata returned by `sbov`</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sbov_output_HS.html">sbov_output_HS</a></code> </p>
        </td>
        <td><p>a GRanges instance with DnaseI hotspot metadata returned by `sbov`</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sbov_output_eQTL.html">sbov_output_eQTL</a></code> </p>
        </td>
        <td><p>a GRanges instance with eQTL metadata returned by `sbov`</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sbov_to_graphNEL.html">sbov_to_graphNEL()</a></code> </p>
        </td>
        <td><p>convert sbov() output to annotated graphNEL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tnt_genetrack_hg19.html">tnt_genetrack_hg19</a></code> </p>
        </td>
        <td><p>gene addresses for visualization with TnTGenome</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tnt_txtrack_hg19.html">tnt_txtrack_hg19</a></code> </p>
        </td>
        <td><p>transcript addresses for visualization with TnTGenome</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="txmodels.html">txmodels()</a></code> </p>
        </td>
        <td><p>use Gviz to render transcript models via GeneRegionTrack, but keep lightweight through requireNamespace and suggestion for installation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="verifyHasMongoCmd.html">verifyHasMongoCmd()</a></code> </p>
        </td>
        <td><p>check for existence of 'mongo' command, for db.getCollectionNames etc.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="verifyRunningMongodb.html">verifyRunningMongodb()</a></code> </p>
        </td>
        <td><p>check for accessible local mongodb</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by VJ Carey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


